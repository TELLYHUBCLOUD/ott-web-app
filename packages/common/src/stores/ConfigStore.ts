import { ACCESS_MODEL, OTT_GLOBAL_PLAYER_ID } from '../constants';
import type { AccessModel, Config, IntegrationType } from '../../types/config';
import type { Settings } from '../../types/settings';

import { createStore } from './utils';

type ConfigState = {
  loaded: boolean;
  config: Config;
  accessModel: AccessModel;
  settings: Settings;
  integrationType: IntegrationType | null;
  isSandbox: boolean;
  clientId: string | null;
  offers: string[];
};

export const useConfigStore = createStore<ConfigState>('ConfigStore', () => ({
  loaded: false,
  config: {
    id: '',
    siteName: '',
    description: '',
    player: '',
    assets: {},
    content: [],
    menu: [],
    integrations: {
      cleeng: {
        useSandbox: true,
      },
      jwp: {
        clientId: null,
        assetId: null,
        useSandbox: true,
      },
    },
    styling: {
      footerText: '',
    },
  },
  settings: {
    additionalAllowedConfigSources: [],
    playerId: OTT_GLOBAL_PLAYER_ID,
  },
  accessModel: ACCESS_MODEL.SVOD,
  integrationType: null,
  isSandbox: true,
  clientId: null,
  offers: [],
}));
