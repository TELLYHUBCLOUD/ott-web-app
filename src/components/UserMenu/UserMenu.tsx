import React, { useCallback } from 'react';
import { useNavigate } from 'react-router-dom';
import { useTranslation } from 'react-i18next';
import classNames from 'classnames';

import styles from './UserMenu.module.scss';
import ProfilesMenu from './ProfilesMenu/ProfilesMenu';

import AccountCircle from '#src/icons/AccountCircle';
import Favorite from '#src/icons/Favorite';
import BalanceWallet from '#src/icons/BalanceWallet';
import Exit from '#src/icons/Exit';
import MenuButton from '#components/MenuButton/MenuButton';
import { useSelectProfile } from '#src/hooks/useProfiles';
import ProfileCircle from '#src/icons/ProfileCircle';
import type { AccessModel } from '#types/Config';
import type { Profile } from '#types/account';
import defaultAvatar from '#src/assets/profiles/default_avatar.png';
import AccountController from '#src/stores/AccountController';
import { getModule } from '#src/modules/container';

type ProfileSpecificProps = {
  small?: boolean;
  currentProfile?: Profile;
  profiles?: Profile[];
};

type Props = {
  showPaymentsItem: boolean;
  onClick?: () => void;
  accessModel?: AccessModel;
  profilesEnabled?: boolean;
} & ProfileSpecificProps;

const UserMenu = ({ showPaymentsItem, small = false, onClick, accessModel, currentProfile, profilesEnabled, profiles }: Props) => {
  const accountController = getModule(AccountController);

  const { t } = useTranslation('user');
  const navigate = useNavigate();

  const onLogout = useCallback(async () => {
    if (onClick) {
      onClick();
    }

    await accountController.logout();
    navigate('/', { replace: true });
  }, [onClick, navigate, accountController]);

  return (
    <ul className={styles.menuItems}>
      {accessModel === 'SVOD' && profilesEnabled && <ProfilesComponent currentProfile={currentProfile} small={small} profiles={profiles} />}
      <li className={styles.sectionHeader}>{t('nav.settings')}</li>
      {profilesEnabled && currentProfile && (
        <li>
          <MenuButton
            small={small}
            onClick={onClick}
            to={`/u/my-profile/${currentProfile?.id ?? ''}`}
            label={t('nav.profile')}
            startIcon={<ProfileCircle src={currentProfile?.avatar_url || defaultAvatar} alt={currentProfile?.name ?? ''} />}
          />
        </li>
      )}
      <li>
        <MenuButton small={small} onClick={onClick} to="/u/my-account" label={t('nav.account')} startIcon={<AccountCircle />} />
      </li>

      <li>
        <MenuButton small={small} onClick={onClick} to="/u/favorites" label={t('nav.favorites')} startIcon={<Favorite />} />
      </li>
      {showPaymentsItem && (
        <li>
          <MenuButton small={small} onClick={onClick} to="/u/payments" label={t('nav.payments')} startIcon={<BalanceWallet />} />
        </li>
      )}
      <hr className={classNames(styles.divider, { [styles.small]: small })} />
      <li>
        <MenuButton small={small} onClick={onLogout} label={t('nav.logout')} startIcon={<Exit />} />
      </li>
    </ul>
  );
};

const ProfilesComponent = ({ profiles, currentProfile, small }: ProfileSpecificProps) => {
  const { t } = useTranslation('user');
  const navigate = useNavigate();
  const selectProfile = useSelectProfile();

  return (
    <ProfilesMenu
      profiles={profiles ?? []}
      currentProfile={currentProfile}
      small={small}
      selectingProfile={selectProfile.isLoading}
      selectProfile={selectProfile.mutate}
      defaultAvatar={defaultAvatar}
      createButtonLabel={t('nav.add_profile')}
      switchProfilesLabel={t('nav.switch_profiles')}
      onCreateButtonClick={() => navigate('/u/profiles/create')}
    />
  );
};

export default UserMenu;
