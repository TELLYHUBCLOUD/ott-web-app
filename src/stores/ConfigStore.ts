import { createStore } from './utils';

import type { AccessModel, Config } from '#types/Config';
import { API_HOST, DEFAULT_PLAYER } from '#src/config';

export enum PersonalShelf {
  ContinueWatching = 'continue_watching',
  Favorites = 'favorites',
}

export const PersonalShelves = [PersonalShelf.Favorites, PersonalShelf.ContinueWatching];

type CleengData = {
  cleengId: string | null | undefined;
  cleengSandbox: boolean;
  monthlyOfferId: string;
  yearlyOfferId: string;
};

type ConfigState = {
  config: Config;
  accessModel: AccessModel;
  apiHost: string;
  getPlayer: () => string;
  getCleengData: () => CleengData;
};

export const useConfigStore = createStore<ConfigState>('ConfigStore', (_, get) => ({
  config: {
    id: '',
    siteName: '',
    description: '',
    player: '',
    assets: {},
    content: [],
    menu: [],
    integrations: {
      cleeng: {
        useSandbox: true,
      },
      inplayer: {
        clientId: null,
        assetId: null,
        useSandbox: true,
      },
    },
    styling: {
      footerText: '',
    },
  },
  accessModel: 'SVOD',
  apiHost: import.meta.env.APP_API_BASE_URL || API_HOST.prd,
  getPlayer: (): string => {
    const host = get().apiHost;

    // For dev api host we use only one testing id, for prod it can be also set in .env file
    return host === API_HOST.dev ? DEFAULT_PLAYER.dev : import.meta.env.APP_DEFAULT_PLAYER ?? DEFAULT_PLAYER.prd;
  },
  getCleengData: (): CleengData => {
    const cleeng = get().config?.integrations?.cleeng;

    const cleengId = cleeng?.id;
    const cleengSandbox = !!cleeng?.useSandbox;
    const monthlyOfferId = cleeng?.monthlyOffer || '';
    const yearlyOfferId = cleeng?.yearlyOffer || '';

    return { cleengId, cleengSandbox, monthlyOfferId, yearlyOfferId };
  },
}));
