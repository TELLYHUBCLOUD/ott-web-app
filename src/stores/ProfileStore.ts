import type { Profile } from '#types/account';
import { createStore } from '#src/utils/createStore';

type ProfileStore = {
  profile: Profile | null;
  selectingProfileAvatar: string | null;
};

export const useProfileStore = createStore<ProfileStore>('AccountStore', () => ({
  profile: null,
  selectingProfileAvatar: null,
}));
