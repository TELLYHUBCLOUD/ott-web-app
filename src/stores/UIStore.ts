import type { Location } from 'react-router-dom';

import { createStore } from '#src/utils/createStore';

type UIState = {
  searchQuery: string;
  searchActive: boolean;
  userMenuOpen: boolean;
  languageMenuOpen: boolean;
  preSearchPage?: Location;
};

export const useUIStore = createStore<UIState>('UIStore', () => ({
  searchQuery: '',
  searchActive: false,
  languageMenuOpen: false,
  userMenuOpen: false,
}));
