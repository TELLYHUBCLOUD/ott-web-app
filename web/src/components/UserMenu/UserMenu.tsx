import React, { useCallback } from 'react';
import { useNavigate } from 'react-router-dom';
import { useTranslation } from 'react-i18next';
import classNames from 'classnames';
import type { Profile } from '@jwplayer/ott-common/types/account';
import { getModule } from '@jwplayer/ott-common/src/modules/container';
import AccountController from '@jwplayer/ott-common/src/stores/AccountController';

import AccountCircle from '../../icons/AccountCircle';
import Favorite from '../../icons/Favorite';
import BalanceWallet from '../../icons/BalanceWallet';
import Exit from '../../icons/Exit';
import MenuButton from '../MenuButton/MenuButton';
import ProfileCircle from '../../icons/ProfileCircle';

import ProfilesMenu from './ProfilesMenu/ProfilesMenu';
import styles from './UserMenu.module.scss';

type Props = {
  small?: boolean;
  showPaymentsItem: boolean;
  onClick?: () => void;
  currentProfile?: Profile | null;
  profilesEnabled?: boolean;
  profiles?: Profile[];
  isSelectingProfile?: boolean;
  selectProfile?: ({ id, avatarUrl }: { id: string; avatarUrl: string }) => void;
  favoritesEnabled?: boolean;
};

const UserMenu = ({
  showPaymentsItem,
  small = false,
  onClick,
  currentProfile,
  profilesEnabled,
  profiles,
  isSelectingProfile,
  selectProfile,
  favoritesEnabled,
}: Props) => {
  const { t } = useTranslation('user');
  const navigate = useNavigate();
  const accountController = getModule(AccountController);

  const onLogout = useCallback(async () => {
    if (onClick) {
      onClick();
    }

    await accountController.logout();
    navigate('/', { replace: true });
  }, [onClick, navigate, accountController]);

  return (
    <ul className={styles.menuItems}>
      {profilesEnabled && selectProfile && (
        <ProfilesMenu
          profiles={profiles ?? []}
          currentProfile={currentProfile}
          small={small}
          selectingProfile={!!isSelectingProfile}
          selectProfile={selectProfile}
          createButtonLabel={t('nav.add_profile')}
          switchProfilesLabel={t('nav.switch_profiles')}
          onCreateButtonClick={() => navigate('/u/profiles/create')}
        />
      )}
      <li className={styles.sectionHeader}>{t('nav.settings')}</li>
      {profilesEnabled && currentProfile && (
        <li>
          <MenuButton
            small={small}
            onClick={onClick}
            to={`/u/my-profile/${currentProfile?.id ?? ''}`}
            label={t('nav.profile')}
            startIcon={<ProfileCircle src={currentProfile?.avatar_url} alt={currentProfile?.name ?? ''} />}
          />
        </li>
      )}
      <li>
        <MenuButton small={small} onClick={onClick} to="/u/my-account" label={t('nav.account')} startIcon={<AccountCircle />} />
      </li>

      {favoritesEnabled && (
        <li>
          <MenuButton small={small} onClick={onClick} to="/u/favorites" label={t('nav.favorites')} startIcon={<Favorite />} />
        </li>
      )}
      {showPaymentsItem && (
        <li>
          <MenuButton small={small} onClick={onClick} to="/u/payments" label={t('nav.payments')} startIcon={<BalanceWallet />} />
        </li>
      )}
      <li>
        <hr className={classNames(styles.divider, { [styles.small]: small })} />
      </li>
      <li>
        <MenuButton small={small} onClick={onLogout} label={t('nav.logout')} startIcon={<Exit />} />
      </li>
    </ul>
  );
};

export default UserMenu;
