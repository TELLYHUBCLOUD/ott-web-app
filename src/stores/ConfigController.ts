import { useConfigStore } from './ConfigStore';

import { API_HOST } from '#src/config';
import { IS_PROD_MODE } from '#src/utils/common';

const envQueryKey = 'env';

export const initializeHost = (searchParams: URLSearchParams) => {
  const query = searchParams.get(envQueryKey);

  if (IS_PROD_MODE) {
    return;
  }

  if (query === 'dev' || query === 'prd') {
    useConfigStore.setState({ apiHost: API_HOST[query] });
    return;
  }

  useConfigStore.setState({ apiHost: API_HOST.prd });
};
