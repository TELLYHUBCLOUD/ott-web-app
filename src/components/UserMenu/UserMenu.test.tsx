import React from 'react';

import { render } from '../../testUtils';
import { ConfigStore } from '../../stores/ConfigStore';

import UserMenu from './UserMenu';

describe('<UserMenu>', () => {
  test('renders and matches snapshot', () => {
    ConfigStore.update((s) => {
      s.config.sso = { host: 'https://www.aws.com', clientId: '12345CLIENT', signingService: 'localhost' };
    });

    const { container } = render(<UserMenu showPaymentsItem={true} />);

    expect(container).toMatchSnapshot();
  });
});
