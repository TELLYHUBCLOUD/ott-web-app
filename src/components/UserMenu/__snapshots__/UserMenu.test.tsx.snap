// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<UserMenu> > renders and matches snapshot 1`] = `
<div>
  <h2>
    Unexpected Application Error!
  </h2>
  <h3
    style="font-style: italic;"
  >
    Cannot destructure property 'authState' of '__vite_ssr_import_5__.useOktaAuth(...)' as it is null.
  </h3>
  <pre
    style="padding: 0.5rem; background-color: rgba(200, 200, 200, 0.5);"
  >
    TypeError: Cannot destructure property 'authState' of '__vite_ssr_import_5__.useOktaAuth(...)' as it is null.
    at UserMenu (/Users/kire.mitrov/projects/jw/ott-web-app/src/components/UserMenu/UserMenu.tsx:35:11)
    at renderWithHooks (/Users/kire.mitrov/projects/jw/ott-web-app/node_modules/react-dom/cjs/react-dom.development.js:16305:18)
    at mountIndeterminateComponent (/Users/kire.mitrov/projects/jw/ott-web-app/node_modules/react-dom/cjs/react-dom.development.js:20074:13)
    at beginWork (/Users/kire.mitrov/projects/jw/ott-web-app/node_modules/react-dom/cjs/react-dom.development.js:21587:16)
    at beginWork$1 (/Users/kire.mitrov/projects/jw/ott-web-app/node_modules/react-dom/cjs/react-dom.development.js:27426:14)
    at performUnitOfWork (/Users/kire.mitrov/projects/jw/ott-web-app/node_modules/react-dom/cjs/react-dom.development.js:26560:12)
    at workLoopSync (/Users/kire.mitrov/projects/jw/ott-web-app/node_modules/react-dom/cjs/react-dom.development.js:26466:5)
    at renderRootSync (/Users/kire.mitrov/projects/jw/ott-web-app/node_modules/react-dom/cjs/react-dom.development.js:26434:7)
    at recoverFromConcurrentError (/Users/kire.mitrov/projects/jw/ott-web-app/node_modules/react-dom/cjs/react-dom.development.js:25850:20)
    at performConcurrentWorkOnRoot (/Users/kire.mitrov/projects/jw/ott-web-app/node_modules/react-dom/cjs/react-dom.development.js:25750:22)
  </pre>
  <p>
    ðŸ’¿ Hey developer ðŸ‘‹
  </p>
  <p>
    You can provide a way better UX than this when your app throws errors by providing your own 
    <code
      style="padding: 2px 4px; background-color: rgba(200, 200, 200, 0.5);"
    >
      ErrorBoundary
    </code>
     or
     
    <code
      style="padding: 2px 4px; background-color: rgba(200, 200, 200, 0.5);"
    >
      errorElement
    </code>
     prop on your route.
  </p>
</div>
`;
